code/dataaccess/getdata.q

```diff
@@ -1,10 +1,15 @@
 // high level api functions for data retrieval
 
-
 getdata:{[inputparams]
-  if[.proc.proctype in key inputparams;inputparams:inputparams .proc.proctype];
+  if[any bool:.proc.proctype=key inputparams;
+  // check if itself is striped and if there is a need to reroute queries
+    inputparams:$[(`attributes in cols value inputparams)&all(s:`skeysym`skeytime)in key att:.proc.getattributes[];
+      [inputparams:value[inputparams]where bool;
+      inputparams first where att[s]~/:inputparams`attributes];
+      inputparams .proc.proctype]];
```

Comment:
Yes this is for the special handling for striping.

The key of the input params have to be of type 11h (proctype). Previously, when there is no striping, `inputparams .proc.proctype` works as it will get the first row where the key matches. But with striping, the respective procs needs to have awareness of which row to get for the correct query. 

This is the `inputparams`:
`   | tablename starttime                     endtime                       columns                                  getquery procs checksperformed attributes                   instruments
---| -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
rdb| quote     2022.01.13D00:00:00.000000000 2022.01.13D02:32:06.805022000 time sym bid ask bsize asize mode ex src 1        rdb   1               0  00:00:00.000 23:59:59.999 \`DELL
rdb| quote     2022.01.13D00:00:00.000000000 2022.01.13D02:32:06.805022000 time sym bid ask bsize asize mode ex src 1        rdb   1               1  00:00:00.000 23:59:59.999 \`DOW
rdb| quote     2022.01.13D00:00:00.000000000 2022.01.13D02:32:06.805022000 time sym bid ask bsize asize mode ex src 1        rdb   1               2  00:00:00.000 23:59:59.999 \`GOOG
hdb| quote     2022.01.12D02:32:06.805022000 2022.01.12D23:59:59.999999999 time sym bid ask bsize asize mode ex src 1        hdb   1               0N \`time$()                  \`GOOG\`DELL\`DOW
`

The first check:
1) `all(s:`\``skeysym`\``skeytime)in key att:.proc.getattributes[]`: checks if itself is striped (its own reporting),
then
2) `(`\``attributes in cols value inputparams)`: checks if there is a need to reroute queries. This is also required because for e.g. if instruments is not specified in the querydict (there will not be an attributes col in the `inputparams`) and the rdb processes are fully striped. Then all the rdb query (row) will be the same, so `inputparams .proc.proctype` will suffice. But each rdbs will get different results as they are reporting different stripe.

`.proc.proctype in key inputparams;inputparams:inputparams .proc.proctype` and `any bool:.proc.proctype=key inputparams` technically gives the same result, but I want to use `bool` to get all the rows instead of just the first (not sure if there is a better way to do this).

Then `first where att[s]~/:inputparams`\``attributes` gets the exact row of query that it should run. So for the above `inputparams` e.g. the query would be 3 diff rdbs querying their respective instruments of \`DELL\`DOW\`GOOG. Instead of 3 rdbs all querying for \`DELL (first row) which then the last 2 rdbs will return an empty result because it does not contain \`DELL.

Let me know if that makes sense!

Reply:
Hmm, OK I think I understand (markdown formatting isn't entirely clear here - for q structures usually better to just paste the structure into a code block instead). I'm a bit wary of having a dictionary with duplicate keys - technically kdb doesn't prevent us doing that, but it can lead to complications like the slightly convoluted indexing here etc. It's also easy to overlook code that implicitly assumes each key appears only once which could cause things to break in unexpected ways.

Would it be possible to re-structure `inputparams` in a different way? Whether that's a new key that gets used in the striped case or something else entirely (e.g. a table or something).

Realise this could end up being a more "invasive" change, but would hopefully reduce the complexity of this code.

Reply:
Sure I will try to implement the changes. 
My idea would be to rename the striped procs to their procname. E.g. stripe key for rdb1 is 0. So if it is striped the key would be \`rdb{skey+1}. So the procs can just have a check if their `.proc.procname` is there. If not it can look for `.proc.proctype`. This matches with what we had in the striping functions we had in the STP.

I will try to keep the same structure so that it does not break any other part of the code.

Reply:
I have made the changes in 9c342a6ab61ef5c37b070ab06491946c3fd4f382. It seems to be working for now. But I will have to do more testing on this.